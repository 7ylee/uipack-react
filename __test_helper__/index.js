import React from 'react';
import { ThemeProvider } from 'styled-components';
import renderer from 'react-test-renderer';
import { idus } from '../src/theme';

export function renderWithTheme(component) {
    return renderer.create(
        <ThemeProvider theme={idus}>
            {component}
        </ThemeProvider>
    );
};

export function foo() {
};