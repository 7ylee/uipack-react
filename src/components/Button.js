import styled, { css } from 'styled-components';
import Color from 'color';
import { resetCSS, boxShadow } from '../utility';
import { normal } from '../theme';

const Button = styled.button`
    ${resetCSS + boxShadow};

    position: relative;
    transition: all 0.2s cubic-bezier(0.175, 0.885, 0.32, 1.275);
    text-decoration: none;
    cursor: pointer;

    /* default element attributes */
    &[disabled] {
        color: ${props => props.theme.GRAY_LIGHT_3000};
        border-color: ${props => props.theme.GRAY_LIGHT_2000};
        background: ${props => props.theme.GRAY_LIGHT_1000};
        cursor: default;
        box-shadow:none;
    };

    /* display props */
    vertical-align: middle;
    display: ${props => props.block ? 'block' : 'inline-block'};

    /* sizing props */
    ${props => {
        if (props.large) {
            return css`
                padding: 0 24px;
                font-size: 1em;
                line-height: 46px;
            `;
        };
        if (props.small) {
            return css`
                padding: 0 12px;
                font-size: 0.6em;
                line-height: 22px;
            `;
        };
        return css`
            padding: 0 16px;
            font-size: 0.8em;
            line-height: 30px;
        `;
    }};

    /* style properties */
    ${props => {
        let propColor = props.theme.WHITE;
        if (props.primary) {
            propColor = props.theme.PRIMARY;
        };
        if (props.secondary) {
            propColor = props.theme.SECONDARY;
        };
        if (props.positive) {
            propColor = props.theme.POSITIVE;
        };
        if (props.negative) {
            propColor = props.theme.NEGATIVE;
        };

        return css`
            border-radius: 2px;
            border-width: 1px;
            border-style: solid;
            border-color: ${Color(propColor).darken(0.3).hex()};
            background-color: ${propColor};
            color: ${Color(propColor).isDark() ? props.theme.WHITE : props.theme.GRAY_3};

            &:focus:not(:disabled),
            &:hover:not(:disabled) {
                outline: none;
                background-color: ${Color(propColor).isDark() ? Color(propColor).darken(0.15).hex() : Color(propColor).darken(0.08).hex()};
            };
        `;
    }};
`;

Button.defaultProps = { theme: {} };
// map normal theme as default;
Object.keys(normal).map(key => {
    Button.defaultProps.theme[key] = normal[key];
    return false
});

export default Button;