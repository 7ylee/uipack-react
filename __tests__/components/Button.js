import React from 'react';
import Button from '../../src/components/Button';
import { renderWithTheme } from '../../__test_helper__';

describe('Button', () => {
    // render with theme
    it('renders correctly with theme', () => {
        const tree = renderWithTheme(
            <Button type="button">Button</Button>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    // style modifiers - has primary, secondary, positive, negative variations
    it('has style modifier - primary', () => {
        const tree = renderWithTheme(
            <Button type="button" primary>Button</Button>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    // style modifiers - has small, large
    it('has size modifier - small', () => {
        const tree = renderWithTheme(
            <Button type="button" small>Button</Button>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    // display modifiers - inline by default
    it('has display modifier - block', () => {
        const tree = renderWithTheme(
            <Button type="button" block>Button</Button>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    // size modifiers - has large, small, medium by default
    it('has size & style & display modifier', () => {
        const tree = renderWithTheme(
            <Button type="button" primary large block>Button</Button>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    // disabled attribute
    it('has disabled attribute', () => {
        const tree = renderWithTheme(
            <Button type="button" primary large block disabled>Button</Button>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });
});