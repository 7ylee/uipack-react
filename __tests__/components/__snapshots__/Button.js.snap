// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button has disabled attribute 1`] = `
<button
  className="sc-bdVaJa cjRWsL"
  disabled={true}
  type="button"
>
  Button
</button>
`;

exports[`Button has display modifier - block 1`] = `
<button
  className="sc-bdVaJa fIZFGC"
  type="button"
>
  Button
</button>
`;

exports[`Button has size & style & display modifier 1`] = `
<button
  className="sc-bdVaJa cjRWsL"
  type="button"
>
  Button
</button>
`;

exports[`Button has size modifier - small 1`] = `
<button
  className="sc-bdVaJa bJXexq"
  type="button"
>
  Button
</button>
`;

exports[`Button has style modifier - primary 1`] = `
<button
  className="sc-bdVaJa gGebTM"
  type="button"
>
  Button
</button>
`;

exports[`Button renders correctly with theme 1`] = `
<button
  className="sc-bdVaJa hPVGnU"
  type="button"
>
  Button
</button>
`;
